Add ledger scans for viewing and freezing keys

shenkeyao opened the pull request:
- Adds scans for viewing and freezing.
- Fixes missed scans in `Keystore::New_impl`.

Closes https://github.com/EspressoSystems/seahorse/issues/270.

bfish713 commented on src/lib.rs:
```diff
@@ -1074,84 +1106,200 @@ impl<
-                .update(|KeystoreSharedState { state, model, .. }| {
-                    state.add_viewing_account(model, Some(viewing_key), description)
-                })
+                .update(
+                    |KeystoreSharedState {
+                         state,
+                         model,
+                         pending_viewing_key_scans,
+                         ..
+                     }| async move {
+                        let (viewing_key, events) = state
+                            .add_viewing_account(
+                                model,
+                                Some(viewing_key),
+                                description,
+                                Some(scan_from),
+                            )
+                            .await?;
+                        // Register the key scan in `pending_viewing_key_scans` so that
+                        // `await_viewing_key_scan` will work.
+                        pending_viewing_key_scans.insert(viewing_key.pub_key(), vec![]);
+                        Ok((viewing_key, events))
+                    },
+                )
                 .await?;
+
+            if let Some(events) = events {
+                // Start a background task to scan for records viewable by the new key.
+                self.spawn_viewing_key_scan(viewing_key.pub_key(), events)
+                    .await;
+            }
+
```
Is it possible to remove the duplicated code at least per key type?

shenkeyao replied:
I tried using two enums `KeyType` and `PubKeyType` to avoid duplications for different `KeyPair` and `PubKey` types, respectively. `KeyType` simplified the relevant code a lot, but `PubkeyType` resulted in much more code which didn't seem that helpful, so I kept `KeyType` but not `PubkeyType`. I'll see if I can simplify this part though.

shenkeyao replied:
Oops, I just realized that the inputs of these functions are _KeyPair_, rather than _PubKey_, so I can definitely simplify this using the `KeyType` enum.

bfish713 commented on src/key_scan.rs:
```diff
@@ -89,31 +144,40 @@ pub enum ScanStatus<L: Ledger> {
+    // These records are kept in a separate pool until the scan is complete so that if the scan
+    // encounters an event which spends some of these records, we can remove the spent records
+    // without ever reflecting them in the keystore's balance.
     records: HashMap<Nullifier, (RecordOpening, u64)>,
```
If I read the code correctly we actually only add to this if the key in scan is `Sending`

shenkeyao replied:
Yeah, only the scan for sending is added to `records` since we don't have the nullifier for viewing or freezing. I think I should change `records`. It should be `HashMap` only for sending, and `HashSet<(RecordOpening, u64)>` for viewing and freezing.

bfish713 commented on src/key_scan.rs:
```diff
@@ -234,32 +332,57 @@ impl<L: Ledger> BackgroundKeyScan<L> {
+                                    if record.asset_def.policy_ref().viewer_pub_key()
+                                        == &viewing_key.pub_key()
+                                    {
+                                        received_records.push(record.clone());
```
I think we also need to push to `self.records` here

shenkeyao replied:
Yes! As I mentioned in another comment, I think `self.records` should be `HashSet` for viewing and freezing, then we can insert `record` to `self.records` without its nullifier.

bfish713 reviewed (commented):
In general I think this make sense but I think it (or me) is missing one part.  Where do we actually add Records for our viewable assets to the records store.  I think we need to do that in order to build freezes/unfreezes on those records.  I probably might need a bit more time to look through things to understand better, but I added comments where I think it's missing.

shenkeyao commented:
> In general I think this make sense but I think it (or me) is missing one part. Where do we actually add Records for our viewable assets to the records store. I think we need to do that in order to build freezes/unfreezes on those records. I probably might need a bit more time to look through things to understand better, but I added comments where I think it's missing.

Thanks @bfish713! One thing that is also missing is that only `update_sending_key_scan` calls `add_records`, which updates the assets, sending accounts, and records stores. I think all the `update_*_key_scan` functions should call `add_records` to update the assets and corresponding accounts stores. We may not be able to update the records store for viewing and freezing though, since we don't have the nullifier.

shenkeyao commented:
> Thanks @bfish713! One thing that is also missing is that only `update_sending_key_scan` calls `add_records`, which updates the assets, sending accounts, and records stores. I think all the `update_*_key_scan` functions should call `add_records` to update the assets and corresponding accounts stores. We may not be able to update the records store for viewing and freezing though, since we don't have the nullifier.

Can we make the `Nullfier` in `Record` optional in order to add viewable/freezable records to the Records store?

bfish713 commented:
> > Thanks @bfish713! One thing that is also missing is that only `update_sending_key_scan` calls `add_records`, which updates the assets, sending accounts, and records stores. I think all the `update_*_key_scan` functions should call `add_records` to update the assets and corresponding accounts stores. We may not be able to update the records store for viewing and freezing though, since we don't have the nullifier.
> 
> Can we make the `Nullfier` in `Record` optional in order to add viewable/freezable records to the Records store?

We could make it option if we need to, but I think we can use the same code we use when we add the record from a ledger event in `receive_attached_records`.  The code is:

```
            // Add the record.
            stores.records.create::<L>(
                *uid,
                ro.clone(),
                account
                    .key()
                    .nullify(&ro.pub_key.address(), *uid, &RecordCommitment::from(&ro)),
            )?;
```

shenkeyao commented:
> > Can we make the `Nullfier` in `Record` optional in order to add viewable/freezable records to the Records store?
> 
> We could make it option if we need to, but I think we can use the same code we use when we add the record from a ledger event in `receive_attached_records`. The code is:
> 
> ```
>             // Add the record.
>             stores.records.create::<L>(
>                 *uid,
>                 ro.clone(),
>                 account
>                     .key()
>                     .nullify(&ro.pub_key.address(), *uid, &RecordCommitment::from(&ro)),
>             )?;
> ```

Sounds good for freezing keys! With viewing keys, I don't think there's a way to get the nullifier so we still make `Nullifier` optional, right?

bfish713 commented:
> > > Can we make the `Nullfier` in `Record` optional in order to add viewable/freezable records to the Records store?
> > 
> > 
> > We could make it option if we need to, but I think we can use the same code we use when we add the record from a ledger event in `receive_attached_records`. The code is:
> > ```
> >             // Add the record.
> >             stores.records.create::<L>(
> >                 *uid,
> >                 ro.clone(),
> >                 account
> >                     .key()
> >                     .nullify(&ro.pub_key.address(), *uid, &RecordCommitment::from(&ro)),
> >             )?;
> > ```
> 
> Sounds good for freezing keys! With viewing keys, I don't think there's a way to get the nullifier so we still make `Nullifier` optional, right?

Oh right I forgot about the viewing part.  Yeah I think making `Nullifier` optional for the viewing keys makes sense.

jbearer commented:
> Sounds good for freezing keys! With viewing keys, I don't think there's a way to get the nullifier so we still make Nullifier optional, right?

This is why we currently only view records that are also freezable. The solution is #6 , essentially, have a new model type for viewing data (which includes something like a record opening whose fields are all optional, but also transaction information and credentials). I think this requires more design (@bfish713 may have some thoughts on this) but assuming the design of that is still not ready to pull the trigger on, I think it's best to just keep the existing limitation (only view when you can also freeze, and then you can get the nullifier)

bfish713 commented:
> > Sounds good for freezing keys! With viewing keys, I don't think there's a way to get the nullifier so we still make Nullifier optional, right?
> 
> This is why we currently only view records that are also freezable. The solution is #6 , essentially, have a new model type for viewing data (which includes something like a record opening whose fields are all optional, but also transaction information and credentials). I think this requires more design (@bfish713 may have some thoughts on this) but assuming the design of that is still not ready to pull the trigger on, I think it's best to just keep the existing limitation (only view when you can also freeze, and then you can get the nullifier)

I agree we can just keep things as they are an not store the viewed but not freezable records until we have a proper place store them.  For now the viewed records would just take up space in the Records model.

shenkeyao commented:
> > > Sounds good for freezing keys! With viewing keys, I don't think there's a way to get the nullifier so we still make Nullifier optional, right?
> > 
> > 
> > This is why we currently only view records that are also freezable. The solution is #6 , essentially, have a new model type for viewing data (which includes something like a record opening whose fields are all optional, but also transaction information and credentials). I think this requires more design (@bfish713 may have some thoughts on this) but assuming the design of that is still not ready to pull the trigger on, I think it's best to just keep the existing limitation (only view when you can also freeze, and then you can get the nullifier)
> 
> I agree we can just keep things as they are an not store the viewed but not freezable records until we have a proper place store them. For now the viewed records would just take up space in the Records model.

Sounds good! Yeah when I was addressing the comments, I realized that it’s fine for now to only include freezable or spendable records, so I kept the nullifier as required. I’ll send an update soon.
